### PR TITLE
feat: add fn body to `process_slashings`

### PR DESCRIPTION
**Pull request by buchhlz2**

### CODE REVIEW

File: src/phase0/state_transition/epoch_processing.rs [hunk 1]
```diff
@@ -177,8 +178,29 @@ pub fn process_slashings<
...
-    _context: &Context,
+    context: &Context,
 ) -> Result<(), Error> {
+    let epoch = get_current_epoch(state, context);
+    let total_balance = get_total_active_balance(state, context)?;
+    let adjusted_total_slashing_balance = Gwei::min(
+        state.slashings.iter().sum::<Gwei>() * context.proportional_slashing_multiplier,
+        total_balance,
+    );
+    let validators = state.validators.clone();
```
Comment (buchhlz2) [inline src/phase0/state_transition/epoch_processing.rs:189]: So, I tried to avoid using `clone()`, but I ran into issues if there's any guidance on better alternatives. 

For example:
```
let validators = state.validators.iter().enumerate();
for (index, validator) in validators {
   ...
   decrease_balance(state, index, penalty);
}
```
When I try to iterate through the data, I get a _"cannot borrow `*state` as mutable because it is also borrowed as immutable"_ error. My understanding is that when this local `validators` variable is created, it borrows state as immutable; thus, when the `for` loop tries to call `decrease_balance` with `state`, it's trying to mutate it -- so, Rust throws an error. 

Basically, since `state.validators` won't be altered during the loop, I'd like to be able to just read it via a reference or something while `decrease_balance` is mutating `state.balances` if that makes sense. I briefly looked into memory-related things like `Cell`, `RefCell`, `RwLock`, `Arc`, etc., but idk if those are needed or how to use them properly.

Comment (ralexstokes) [reply src/phase0/state_transition/epoch_processing.rs:189]: right, this comes down to how rust handles borrows, so we will have to be a little careful about how we do this.

when you say `state.validators.iter()`, you get back a thing that essentially has an immutable reference inside it (generated from the "broader" mutable reference from `state: &mut T`)

and the borrow rule is basically "1 writer xor many readers" -- so if we hand out the immutable ref, we cannot use the mutable ref down below in `decrease_balance` as the immutable ref that is live stays live until the end of its enclosing scope.

there are a few ways to handle this particular problem (e.g. you could make a separate list of the balance deltas to apply and then basically drop the validators ref and *then* apply the deltas) -- one that i think sticks closely to the spec is to just iterate over indices and take advantage of the fact that the `validator` ref is not used in `decrease_balance`

refer to this block: https://github.com/ralexstokes/ethereum_consensus/blob/main/src/phase0/state_transition/genesis.rs#L87-L99

for an example, although you should be able to just say `let validator = &state.validators[i];`. this ends up being fine as this ref for the validator is dropped before we "reborrow" the mutable ref for state in `decrease_balance`

Comment (ralexstokes) [reply src/phase0/state_transition/epoch_processing.rs:189]: as a side note, the `validators` field is one of the larger ones in the state so we definitely don't want to be copying it where we can avoid it

**Review by ralexstokes — COMMENTED** [review]
nice work!

left some comments below

File: src/phase0/state_transition/epoch_processing.rs (hunk 1 above)
Comment (buchhlz2) [reply src/phase0/state_transition/epoch_processing.rs]: Ahh okay, I think I get it now. By using `validators` only within the `for` loop's scope (and not prior to the loop executing), the immutable reference to `state` is last used prior to `decrease_balance` borrowing `state` as mutable...so, then there's no issue with multiple immutable/mutable references at the same time.

**Comment by buchhlz2** [comment]
Oops, didn't see the other comments....will update those too.

File: src/phase0/state_transition/epoch_processing.rs [hunk 2]
```diff
@@ -177,8 +178,29 @@ pub fn process_slashings<
         MAX_VALIDATORS_PER_COMMITTEE,
         PENDING_ATTESTATIONS_BOUND,
     >,
-    _context: &Context,
+    context: &Context,
 ) -> Result<(), Error> {
+    let epoch = get_current_epoch(state, context);
+    let total_balance = get_total_active_balance(state, context)?;
+    let adjusted_total_slashing_balance = Gwei::min(
+        state.slashings.iter().sum::<Gwei>() * context.proportional_slashing_multiplier,
```
[the remark this answers is not in the document]
Comment (buchhlz2) [reply src/phase0/state_transition/epoch_processing.rs:187]: Yeah idk why I needed that, but if I remove it, I get an error:

> cannot satisfy `_: std::iter::Sum<&u64>`
help: consider specifying the type argument in the method call

File: src/phase0/state_transition/epoch_processing.rs [hunk 3]
```diff
@@ -177,8 +178,29 @@ pub fn process_slashings<
...
+
+    for (index, validator) in validators.iter().enumerate() {
+        if validator.slashed
+            && (epoch + context.epochs_per_slashings_vector as u64 / 2)
+                == validator.withdrawable_epoch
+        {
+            let increment = context.effective_balance_increment;
+            let penalty_numerator =
+                validator.effective_balance / increment * adjusted_total_slashing_balance;
+            let penalty = penalty_numerator / total_balance * increment;
```
[the remark this answers is not in the document]
Comment (buchhlz2) [reply src/phase0/state_transition/epoch_processing.rs:200]: Haven't had to think about language-specific PEMDAS in _quite_ some time lol. Alright so python puts (floor) division & multiplication on the same tier. And Lighthouse calls `.safe_div(total_balance)` followed by `.safe_mul(increment)` on `penalty_numerator` within their implementation of this function...so I think what I originally had is correct?

File: src/phase0/state_transition/epoch_processing.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ralexstokes) [reply src/phase0/state_transition/epoch_processing.rs:187]: i think we can just get away with `state.slashings.iter().copied().sum()`

the error suggests it doesn't like the `&Gwei` from the plain `iter` and i think `into_iter` will fail bc it would try to consume the leading ref

usually you could reach for `cloned` in this situation but bc we know our type is `Copy` we will use `copied` (see answers for reasoning: https://stackoverflow.com/questions/58413583/difference-between-copied-and-cloned-on-rust-iterators)

File: src/phase0/state_transition/epoch_processing.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ralexstokes) [reply src/phase0/state_transition/epoch_processing.rs:200]: yep sounds good then, thanks for checking :)

File: src/phase0/state_transition/epoch_processing.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (buchhlz2) [reply src/phase0/state_transition/epoch_processing.rs:187]: Hmm so none of those solved it either -- what [I'm reading](https://github.com/rust-lang/rust/issues/25094) is that `sum()` is super picky with type inferences and needs to have a type specified. For example, when I remove the multiplication by `context.proportional_slashing_multiplier`, it doesn't complain & compiles:
```
let adjusted_total_slashing_balance = Gwei::min(state.slashings.iter().sum(), total_balance);
```
...which is sort of confusing as well since `total_balance` is the same type as `context.proportional_slashing_multiplier`, so it seems like it gets super picky once operations are attempted.

[the remark this answers is not in the document]
Comment (ralexstokes) [reply src/phase0/state_transition/epoch_processing.rs:187]: hmm interesting, well thanks for looking into it!

**Review by ralexstokes — APPROVED** [review]
lgtm!

File: src/phase0/state_transition/epoch_processing.rs [hunk 4]
```diff
@@ -178,8 +179,29 @@ pub fn process_slashings<
...
+    let total_balance = get_total_active_balance(state, context)?;
+    let adjusted_total_slashing_balance = Gwei::min(
+        state.slashings.iter().sum::<Gwei>() * context.proportional_slashing_multiplier,
+        total_balance,
+    );
+
+    for i in 0..state.validators.len() {
+        let validator = &state.validators[i];
+        if validator.slashed
+            && (epoch + context.epochs_per_slashings_vector as u64 / 2)
```
Comment (ralexstokes) [inline src/phase0/state_transition/epoch_processing.rs:194]: making a note to myself patch up the cast as part of #38.
